File: Java/Language/Runtime/FunctionalTest1/Investment.java
//this class produces immutable object whose state cannot be changed
//after instantiation
class Investment {

	//final instance field can only be initialized in the constructor
	private final double payment;
	private final int years;

	public Investment(double payment, int years) {
		this.payment = payment;
		this.years = years;
	}

	public double payment() {
		return payment;
	}

	public int years() {
		return years;
	}

	public double futureValue(InterestRate interest) {
		float i = interest.forPeriod(years) / 100;
		return (payment / i) * (Math.pow(1 + i, years) - 1);
	}
}

File: dotnet/Language/Runtime/DelegateTest/DemoApp/Investment.cs
delegate float InterestRate(int period);

//this class produces an instance which is an immutable object
class Investment
{
    public double Payment { get; }

    public int Years { get; }

    public Investment(double amount, int count)
    {
        Payment = amount;
        Years = count;
    }  

    public double FutureValue(InterestRate rate)
    {
        float i = rate.Invoke(Years) / 100;
        return (Payment / i) * (Math.Pow(1 + i, Years) - 1);
    }
}
